### PR TITLE
build: remove initctl if no sysv support

### PR DESCRIPTION
**Pull request by nolange**
Stil is an atempt to allow an "legacy-free" installation of systemd for embedded systemd. The remains of other init systems are confusing at best.

There are a cople of points I need help with:

- still needs update of systemd/man/rules/meson.build?
- should also remove runlevel*.target sbin/runlevel sbin/telinit symlinks

### CODE REVIEW

**Comment by floppym** [comment]
NAK from me on this. Gentoo Linux wants to install the initctl stuff even when support for sysv init scripts is disabled.

We support installi systemd and OpenRC (on top of sysvinit) concurrently, and will continue to do so for as long as it is feasible. The sysadmin can switch between the two via the kernel command line.

OpenRC init scripts are incompatible with traditional sysvinit scripts, so we disable systemd support for the latter.

**Comment by floppym** [comment]
Maybe add another meson flag for this.

**Comment by nolange** [comment]
Ain't that a packaging thing? On Debian, you build with all options enabled, and if you want openrc, you could pack initctl in the package providing that.
What are your thought of all the runlevel symlinks?

**Comment by floppym** [comment]
> Ain't that a packaging thing?

The HAVE_SYSV_COMPAT preprocessor macro affects code in several systemd binaries. It must be *disabled* on Gentoo.

Your proposed change makes it also control compilation of `systemd-initctl` and several related units. These must be *enabled* on Gentoo.

It makes no sense to build systemd twice just so we can get the `systemd-initctl` binary. That's impractical on a build-from-source distro like Gentoo anyway.

**Comment by floppym** [comment]
> What are your thought of all the runlevel symlinks?

It think those have been obsolete since d5d8429a12c4b1ef0dcd226c0904f00f4fa4898a.

**Comment by floppym** [comment]
> The HAVE_SYSV_COMPAT preprocessor macro affects code in several systemd binaries. It must be disabled on Gentoo.

Upon further review, it would actually be fine for this code to be enabled on Gentoo. We really just want to discard the sysv-generator binary, which should be simple to do after building.

I want to play with that a bit, but consider my objection lifted for the time being.

**Comment by poettering** [comment]
PR makes sense to me.

The man pages are conditioned out via `conditional=` attributes in the `<refentry>` tag at the top of the XML files.

**Comment by floppym** [comment]
I enabled sysv compat on Gentoo, and applied the patch below to prevent systemctl from trying to sync enablement of init scripts.

https://gitweb.gentoo.org/repo/gentoo.git/tree/sys-apps/systemd/files/gentoo-systemctl-disable-sysv-sync.patch

It would be nice to have a separate option to disable systemd-sysv-generator and the systemctl code affected by the above patch.

**Comment by nolange** [comment]
@floppym Skimming through the sources, it seems to me that is easier if I manually remove systemd-initctl and the related service on my installation.
There are alot implications with upstart and openrc and I have little experience with those.

New commit just removes the (as far as I understand it) unnecessary runlevel/telinit symlinks

**Comment by nolange** [comment]
@floppym Now I am just disabling installation of `systemd-initctl`, should be possibly to install them manually. Seems saner to me than enabling all those `HAVE_SYSV_COMPAT` sections.

**Comment by floppym** [comment]
It seems a little weird to build the stuff and then not install it. I appreciate the gesture, but I don't think this is a good approach.

File: units/meson.build [hunk 1]
```diff
@@ -1,5 +1,7 @@
 # SPDX-License-Identifier: LGPL-2.1+
 
+rlvl = conf.get('HAVE_SYSV_COMPAT') == 1
```
Comment (poettering) [inline units/meson.build]: please, let's not abbreviate this so unnecessarily. There's nothing wrong with calling this `with_runlevels`.

File: units/meson.build [hunk 2]
```diff
@@ -42,26 +44,26 @@ units = [
...
+         (rlvl ? 'runlevel2.target runlevel3.target runlevel4.target' : '')],
         ['network-online.target',               ''],
         ['network-pre.target',                  ''],
         ['network.target',                      ''],
         ['nss-lookup.target',                   ''],
         ['nss-user-lookup.target',              ''],
         ['paths.target',                        ''],
         ['poweroff.target',                     '',
-         'runlevel0.target'],
+         (rlvl ? 'runlevel0.target ' : '')],
```
Comment (poettering) [inline units/meson.build]: one space too much in the string

File: units/meson.build [hunk 3]
```diff
@@ -42,26 +44,26 @@ units = [
...
         ['proc-sys-fs-binfmt_misc.mount',       'ENABLE_BINFMT'],
         ['reboot.target',                       '',
-         'runlevel6.target ctrl-alt-del.target'],
+         (rlvl ? 'runlevel6.target ' : '') + 'ctrl-alt-del.target'],
         ['remote-cryptsetup.target',            'HAVE_LIBCRYPTSETUP'],
         ['remote-fs-pre.target',                ''],
         ['remote-fs.target',                    ''],
         ['rescue.target',                       '',
-         'runlevel1.target'],
+         (rlvl ? 'runlevel1.target ' : '')],
```
Comment (poettering) [inline units/meson.build]: ditto

**Comment by nolange** [comment]
Seems the tertiary operator triggers a bug in meson versions < 0.50 (https://github.com/mesonbuild/meson/issues/5003). Modified the code a bit to work around that.

**Comment by keszybz** [comment]
LGTM. We want to keep support on the kernel command line (3, 5, etc.) even if sysvinit compat is disabled. But that is governed by runlevel_to_target() which is independent, so nothing should change with this patch.

**Pull request merged**